core/server/services/auth/session/index.js

```diff
@@ -0,0 +1,130 @@
+const {URL} = require('url'),
+    common = require('../../../lib/common'),
+    config = require('../../../config'),
+    settingsCache = require('../../settings/cache'),
+    {User} = require('../../../models/user'),
+    {Session} = require('../../../models/session'),
+    session = require('express-session'),
+    SessionStore = require('./store'),
+    {BadRequestError, UnauthorizedError, InternalServerError} = require('../../../lib/common/errors');
+
+const getOrigin = function getOrigin(req) {
+    const origin = req.get('origin');
+    const referrer = req.get('referrer');
+
+    if (!origin && !referrer) {
+        return null;
+    }
+
+    if (origin) {
+        return origin;
+    }
+
+    try {
+        return new URL(req.get('referrer')).origin;
+    } catch (e) {
+        return null;
+    }
+};
+
+const createSession = function createSession(req, res, next) {
+    if (!req.body) {
+        return next(new UnauthorizedError({
+            message: common.i18n.t('errors.middleware.auth.accessDenied')
+        }));
+    }
+    const origin = getOrigin(req);
+    if (!origin) {
+        return next(new BadRequestError({
+            message: common.i18n.t('errors.middleware.auth.unknownOrigin')
+        }));
+    }
+    const {username, password} = req.body;
+    User.check({
+        email: username,
+        password
+    }).then((user) => {
+        req.session.user_id = user.id;
+        req.session.origin = origin;
+        req.session.user_agent = req.get('user-agent');
+        req.session.ip = req.ip;
+        res.sendStatus(201);
+    }).catch(() => {
+        next(new UnauthorizedError({
+            message: common.i18n.t('errors.middleware.auth.accessDenied')
+        }));
+    });
+};
+
+const destroySession = function destroySession(req, res, next) {
+    req.session.destroy((err) => {
+        if (err) {
+            return next(new InternalServerError());
```

Comment:
Is this an internal error? 🤔 Internal errors are [marked](https://github.com/TryGhost/Ignition/blob/master/lib/errors/index.js#L24) as **critical** by default. The differentiation between `critical` and `normal` makes it possible to select certain errors to send to a centralised monitoring service e.g. loggly, ES etc.

---

One addition: we try to inherit from a third party error. Otherwise the error stacks would only capture from here on. We also loose the original error message from the third party lib.

`new common.errors. InternalServerError({message: 'lalala', err})`

Reply:
Yeah I would consider this an `internal` error, it means we had an issue deleting from the sessions table, but immediately after reading this object from the sessions table. what do you think?

I'll add the original error in :+1:

Reply:
Does `getSession` throw an error if `req.session` is not defined?

Reply:
No it doesn't - `getSession` _defines_ `req.session` and populates it if there is a valid `Cookie` header

Reply:
Does that mean it's possible that we try to destroy an empty session?
Does express-session has documented error codes or so?

The reason i am asking, i would just like to avoid sending "random" session delete errors to loggly. If there is a critical case, we can detect and log this as critical error.

Reply:
The deleteSession middleware is behind the `auth.auth.adminAPI` middleware - so we will 401 if people try to delete a session when they don't have one